File: hack/integration/ordered_bucket.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// exerciseOrderedBucket is the Ordered Bucket live qualification on the
// explicitly owned disposable cluster, whose third node sits in a second zone.
func (h *harness) exerciseOrderedBucket() {
	alpha := bucketFleet("alpha", "bucket-alpha")
	alpha.Spec.BucketWorkload = "Ordered"
	alpha.Spec.Replicas = 3
	alpha.Spec.Placement.AZCount = 2
	alpha.Spec.Placement.Zones = []string{"us-east-1a", "us-east-1b"}
	h.apply(alpha)
	pods := func() []object { return h.fleetPods("alpha") }
	settled := func(count int64) bool {
		current := h.get("celldfleet", "alpha")
		sts, err := h.tryGet("fleets", "statefulset", "alpha")
		if err != nil {
			if strings.Contains(err.Error(), "(NotFound)") {
				return false
			}
			must(err)
		}
		return specReplicas(sts) == count && num(sts, "status", "readyReplicas") == count && isReady(current) && operationID(current) == ""
	}
	placement := func(count int) {
		current := pods()
		assert(len(current) == count, "%d pods, expected %d", len(current), count)
		names := map[string]bool{}
		for _, pod := range current {
			names[nameOf(pod)] = true
		}
		expected := map[string]bool{}
		for i := range count {
			expected[fmt.Sprintf("alpha-%d", i)] = true
		}
		assert(same(names, expected), "membership %v", names)
		assert(len(nodeNames(current)) == count, "ordered replicas share a node")
		for _, pod := range current {
			ordinal, err := strconv.Atoi(nameOf(pod)[strings.LastIndex(nameOf(pod), "-")+1:])
			must(err)
			zone := []string{"us-east-1a", "us-east-1b"}[ordinal%2]
			node := h.cluster("node", str(pod, "spec", "nodeName"))
			assert(str(node, "metadata", "labels", "topology.kubernetes.io/zone") == zone, "%s scheduled outside %s", nameOf(pod), zone)
			assert(str(pod, "spec", "nodeSelector", "topology.kubernetes.io/zone") == zone, "%s selects the wrong zone", nameOf(pod))
			assert(len(list(pod, "spec", "schedulingGates")) == 0, "%s still gated", nameOf(pod))
		}
		assert(len(h.listIn("fleets", "pvc")) == 0, "Ordered Bucket created claims")
		fmt.Println("PASS: exact ordered membership, AZ mapping, hostname separation, and emptyDir")
	}
	scale := func(count int) {
		h.setReplicas("alpha", count)
		h.waitFor(fmt.Sprintf("Ordered Bucket settled at %d", count), 360*time.Second, func() bool { return settled(int64(count)) })
		placement(count)
	}
	h.waitFor("Ordered Bucket gated pods start across two AZs", 360*time.Second, func() bool { return settled(3) })
	placement(3)
	client := sleepPod("ordered-client", "fleets", curlImage, map[string]string{"celld.eric.dev/client-of": "alpha"})
	client.Spec.Containers[0].Name = "client"
	h.apply(client)
	h.k("-n", "fleets", "wait", "--for=condition=Ready", "pod/ordered-client", "--timeout=90s")
	request := func(method, path string) object {
		return decode(h.k("-n", "fleets", "exec", "ordered-client", "--", "curl", "--fail", "--silent", "--show-error", "--max-time", "10", "-X", method, "http://alpha:8080"+path))
	}
	paths := make([]string, 0, 12)
	for i := range 12 {
		paths = append(paths, fmt.Sprintf("/?cell=ordered&id=ack-%d", i))
	}
	for _, path := range paths {
		assert(stored(request("PUT", path)), "write %s not acknowledged", path)
	}
	original := nameUIDs(pods())
	scale(2)
	expected := map[string]string{}
	for key, value := range original {
		if key != "alpha-2" {
			expected[key] = value
		}
	}
	assert(same(nameUIDs(pods()), expected), "contraction replaced a surviving ordinal")
	for _, path := range paths {
		assert(stored(request("GET", path)), "write %s lost", path)
	}
	fmt.Println("PASS: only highest ordinal removed; all 12 acknowledged writes survived")
	// This node was admitted by completed removal. Restart its unchanged pinned
	// runtime in place, then require positive generation supersession on the next removal.
	before := h.get("pod", "alpha-0")
	oldContainer := str(list(before, "status", "containerStatuses")[0], "containerID")
	h.k("-n", "fleets", "exec", "alpha-0", "--", "/bin/sh", "-c", "kill -TERM 1")
	h.waitFor("same Pod UID acquires a new ready runtime invocation", 180*time.Second, func() bool {
		statuses := list(h.get("pod", "alpha-0"), "status", "containerStatuses")
		return len(statuses) > 0 && str(statuses[0], "containerID") != oldContainer && boolean(statuses[0], "ready")
	})
	assert(uidOf(h.get("pod", "alpha-0")) == uidOf(before), "in-place restart replaced the Pod")
	scale(3)
	assert(uidOf(h.get("pod", "alpha-2")) != original["alpha-2"], "readmitted ordinal reused the old Pod")
	h.restartOperator()
	scale(2)
	for _, path := range paths {
		assert(stored(request("GET", path)), "write %s lost", path)
	}
	fleetUID := uidOf(h.get("celldfleet", "alpha"))
	var journal object
	for _, reservation := range h.reservations() {
		if str(reservation, "spec", "fleetUID") == fleetUID {
			journal = decode(annotation(reservation, journalAnnotation))
		}
	}
	superseded := false
	for _, session := range list(journal, "BucketHistory") {
		if truthy(field(session, "SupersededBy")) {
			superseded = true
		}
	}
	assert(superseded, "no positive successor recorded in %v", list(journal, "BucketHistory"))
	fmt.Println("PASS: positive successor history survives manager restart and repeated contraction; 12/12 ledger intact")
	_, err := h.tryK("-n", "fleets", "exec", "ordered-client", "--", "curl", "--fail", "--silent", "--max-time", "3", "http://"+str(h.get("pod", "alpha-0"), "status", "podIP")+":8081/state")
	assert(err != nil, "application client reached private state endpoint")
	fmt.Println("PASS: private peer/state endpoint remains inaccessible to application clients")
}

func nameUIDs(pods []object) map[string]string {
	out := map[string]string{}
	for _, pod := range pods {
		out[nameOf(pod)] = uidOf(pod)
	}
	return out
}

// truthy mirrors Python truthiness for decoded JSON values.
func truthy(value any) bool {
	switch v := value.(type) {
	case nil:
		return false
	case string:
		return v != ""
	case bool:
		return v
	case float64:
		return v != 0
	case []any:
		return len(v) > 0
	case object:
		return len(v) > 0
	default:
		return true
	}
}

File: hack/integration/ordered_bucket.py
"""Ordered Bucket live qualification on an explicitly owned disposable cluster."""
import json


def exercise(k, apply, get, wait_for, restart_operator, fleet, curl_image):
    alpha = fleet('alpha', 'bucket-alpha')
    alpha['spec'].update({'bucketWorkload': 'Ordered', 'replicas': 3,
                          'placement': {'azCount': 2, 'zones': ['us-east-1a', 'us-east-1b']}})
    apply(alpha)
    def pods():
        uid = get('celldfleet', 'alpha')['metadata']['uid']
        return json.loads(k('-n', 'fleets', 'get', 'pods', '-l', 'celld.example.com/fleet-uid=' + uid, '-o', 'json'))['items']
    def settled(count):
        current = get('celldfleet', 'alpha')
        try:
            sts = get('statefulset', 'alpha')
        except RuntimeError as error:
            if '(NotFound)' in str(error):
                return False
            raise
        return (sts['spec']['replicas'] == count and sts.get('status', {}).get('readyReplicas') == count
                and any(c['type'] == 'Ready' and c['status'] == 'True' for c in current.get('status', {}).get('conditions', []))
                and not current.get('status', {}).get('lifecycle', {}).get('operationID'))
    def placement(count):
        current = pods()
        assert len(current) == count
        assert {p['metadata']['name'] for p in current} == {'alpha-' + str(i) for i in range(count)}
        assert len({p['spec']['nodeName'] for p in current}) == count
        for pod in current:
            ordinal = int(pod['metadata']['name'].rsplit('-', 1)[1])
            zone = ['us-east-1a', 'us-east-1b'][ordinal % 2]
            node = get('node', pod['spec']['nodeName'], '')
            assert node['metadata']['labels']['topology.kubernetes.io/zone'] == zone
            assert pod['spec']['nodeSelector']['topology.kubernetes.io/zone'] == zone
            assert not pod['spec'].get('schedulingGates')
        assert not json.loads(k('-n', 'fleets', 'get', 'pvc', '-o', 'json'))['items']
        print('PASS: exact ordered membership, AZ mapping, hostname separation, and emptyDir', flush=True)
    def scale(count):
        k('-n', 'fleets', 'patch', 'celldfleet', 'alpha', '--type=merge', '-p', json.dumps({'spec': {'replicas': count}}))
        wait_for(lambda: settled(count), 'Ordered Bucket settled at ' + str(count), timeout=360)
        placement(count)
    wait_for(lambda: settled(3), 'Ordered Bucket gated pods start across two AZs', timeout=360)
    placement(3)
    apply({'apiVersion': 'v1', 'kind': 'Pod', 'metadata': {'name': 'ordered-client', 'namespace': 'fleets', 'labels': {'celld.example.com/client-of': 'alpha'}}, 'spec': {'containers': [{'name': 'client', 'image': curl_image, 'command': ['/bin/sh', '-c', 'sleep 3600']}]}})
    k('-n', 'fleets', 'wait', '--for=condition=Ready', 'pod/ordered-client', '--timeout=90s')
    def request(method, path):
        return json.loads(k('-n', 'fleets', 'exec', 'ordered-client', '--', 'curl', '--fail', '--silent', '--show-error', '--max-time', '10', '-X', method, 'http://alpha:8080' + path))
    paths = ['/?cell=ordered&id=ack-' + str(i) for i in range(12)]
    for path in paths:
        assert request('PUT', path)['stored'] is True
    original = {p['metadata']['name']: p['metadata']['uid'] for p in pods()}
    scale(2)
    assert {p['metadata']['name']: p['metadata']['uid'] for p in pods()} == {key: value for key, value in original.items() if key != 'alpha-2'}
    for path in paths:
        assert request('GET', path)['stored'] is True
    print('PASS: only highest ordinal removed; all 12 acknowledged writes survived', flush=True)
    # This node was admitted by completed removal. Restart its unchanged pinned
    # runtime in place, then require positive generation supersession on the next removal.
    before = get('pod', 'alpha-0')
    old_container = before['status']['containerStatuses'][0]['containerID']
    k('-n', 'fleets', 'exec', 'alpha-0', '--', '/bin/sh', '-c', 'kill -TERM 1')
    wait_for(lambda: get('pod', 'alpha-0')['status']['containerStatuses'][0].get('containerID') != old_container
             and get('pod', 'alpha-0')['status']['containerStatuses'][0].get('ready'), 'same Pod UID acquires a new ready runtime invocation', timeout=180)
    assert get('pod', 'alpha-0')['metadata']['uid'] == before['metadata']['uid']
    scale(3)
    assert get('pod', 'alpha-2')['metadata']['uid'] != original['alpha-2']
    restart_operator()
    scale(2)
    for path in paths:
        assert request('GET', path)['stored'] is True
    reservations = json.loads(k('get', 'celldstoragereservations', '-o', 'json'))['items']
    journal = json.loads(next(r for r in reservations if r['spec']['fleetUID'] == get('celldfleet', 'alpha')['metadata']['uid'])['metadata']['annotations']['celld.example.com/lifecycle-journal'])
    assert any(s.get('SupersededBy') for s in journal['BucketHistory'])
    print('PASS: positive successor history survives manager restart and repeated contraction; 12/12 ledger intact', flush=True)
    denied = False
    try:
        k('-n', 'fleets', 'exec', 'ordered-client', '--', 'curl', '--fail', '--silent', '--max-time', '3', 'http://' + get('pod', 'alpha-0')['status']['podIP'] + ':8081/state')
    except RuntimeError:
        denied = True
    assert denied, 'application client reached private state endpoint'
    print('PASS: private peer/state endpoint remains inaccessible to application clients', flush=True)
